contracts/core/BorrowableMarginPool.sol

```diff
@@ -0,0 +1,314 @@
+        require(_oTokenBuyer != address(0), "!_buyer");
+
+        whitelistedOTokenBuyer[_oTokenBuyer] = _whitelisted;
+        emit SetOTokenBuyerWhitelist(_oTokenBuyer, _whitelisted);
+    }
+
+    /**
+     * @notice Set Borrow percent of collateral asset
+     * @param _collateral address of collateral asset
+     * @param _borrowPCT borrow PCT
+     */
+    function setBorrowPCT(address _collateral, uint256 _borrowPCT) external onlyOwner {
+        borrowPCT[_collateral] = _borrowPCT;
+        emit SetBorrowPCT(_collateral, _borrowPCT);
+    }
+
+    /**
+     * @notice Lends out asset to market maker
+     * @param _oToken address of the oToken laying claim to the collateral assets
+     * @param _oTokenAmount amount of the oToken to post as collateral in exchange
+     */
+    function borrow(address _oToken, uint256 _oTokenAmount) public onlyWhitelistedBorrower {
+        require(
+            WhitelistInterface(AddressBookInterface(addressBook).getWhitelist()).isWhitelistedOtoken(_oToken),
+            "MarginPool: oToken is not whitelisted"
+        );
+
+        require(_oTokenAmount > 0, "MarginPool: Cannot borrow 0 of underlying");
+
+        OtokenInterface oToken = OtokenInterface(_oToken);
+
+        address collateralAsset = oToken.collateralAsset();
+        uint256 outstandingAssetBorrow = borrowed[msg.sender][collateralAsset];
+
+        require(!oToken.isPut(), "MarginPool: oToken is not a call option");
+
+        // Make sure borrow does not attempt to borrow with an expired oToken
+        require(
+            oToken.expiryTimestamp() > block.timestamp,
+            "MarginPool: Cannot borrow collateral asset of expired oToken"
+        );
+
+        uint256 oTokenBalance = ERC20Interface(_oToken).balanceOf(msg.sender);
+
+        // Each oToken represents 1:1 of collateral token
+        // So a market maker can borrow at most our oToken balance in the collateral asset
+        uint256 oTokenAmountCustodied = _collateralAssetToOTokenAmount(collateralAsset, outstandingAssetBorrow);
+
+        require(
+            _oTokenAmount <=
+                oTokenBalance.add(oTokenAmountCustodied).mul(borrowPCT[collateralAsset]).div(TOTAL_PCT).sub(
+                    oTokenAmountCustodied
+                ),
+            "MarginPool: Borrowing more than allocated"
+        );
+
+        uint256 collateralAssetAmount = _oTokenToCollateralAssetAmount(collateralAsset, _oTokenAmount);
+
+        borrowed[msg.sender][collateralAsset] = outstandingAssetBorrow.add(collateralAssetAmount);
+
+        // transfer _oTokenAmount of oToken from borrower to _pool
+        ERC20Interface(_oToken).safeTransferFrom(msg.sender, address(this), _oTokenAmount);
+        // transfer collateralAssetAmount of collateralAsset from pool to borrower
+        ERC20Interface(collateralAsset).safeTransfer(msg.sender, collateralAssetAmount);
+        emit Borrow(_oToken, collateralAsset, collateralAssetAmount, msg.sender);
```

Comment:
Maybe use the `transferToUser()` and `transferToPool()` to not mess up the pool accounting? wyt?

Reply:
yeah will add `assetBalance[_asset] = assetBalance[_asset].add(_amount);` and `assetBalance[_asset] = assetBalance[_asset].sub(_amount);`

Reply:
Not going to update asset balance for oToken because it wasn't really intended for that purpose wyt

Reply:
wdym? I think it best to make sure the pool accounting always match the pool balance, even for oToken